src/dmd/semantic3.d

```diff
@@ -293,343 +281,183 @@ private extern(C++) final class Semantic3Visitor : Visitor
-                sym.parent = sc2.scopesym;
-                sym.loc = funcdecl.loc;
-                sym.endlinnum = fensure_endlin;
-                scout = sc2.push(sym);
-            }
+            e = e.ctfeInterpret();
+            ad.getRTInfo = e;
+        }
+        if (sd)
+            sd.semanticTypeInfoMembers();
+        ad.semanticRun = PASS.semantic3done;
+    }
+}
 
-            if (funcdecl.fbody)
-            {
-                auto sym = new ScopeDsymbol();
-                sym.parent = sc2.scopesym;
-                sym.loc = funcdecl.loc;
-                sym.endlinnum = funcdecl.endloc.linnum;
-                sc2 = sc2.push(sym);
+private struct FuncDeclSem
+{
+    // Scopes:
+    // Scope of analysis
+    Scope* sc;
+    // Scope created to avoid name collisions
+    Scope* sc2;
+    // Scope for out contract return variable
+    Scope* scout;
 
-                auto ad2 = funcdecl.isMember2();
+    // The FuncDeclaration subject to Semantic analysis
+    FuncDeclaration funcdecl;
+    TypeFunction f;  //funcdecl.type
+    // Does funcdecl need a closure?
+    immutable bool needEnsure;
 
-                /* If this is a class constructor
-                 */
-                if (ad2 && funcdecl.isCtorDeclaration())
-                {
-                    sc2.ctorflow.allocFieldinit(ad2.fields.dim);
-                    foreach (v; ad2.fields)
-                    {
-                        v.ctorinit = 0;
-                    }
-                }
+    // The variable arguemnt ... parameter.
+    VarDeclaration _arguments = null;
 
-                if (!funcdecl.inferRetType && !Target.isReturnOnStack(f, funcdecl.needThis()))
-                    funcdecl.nrvo_can = 0;
+    // Precondition and postcondition invariants
+    Statement fpreinv, fpostinv;
 
-                bool inferRef = (f.isref && (funcdecl.storage_class & STC.auto_));
+    // In / Out contract statements
+    Statement freq, fens;
 
-                funcdecl.fbody = funcdecl.fbody.statementSemantic(sc2);
-                if (!funcdecl.fbody)
-                    funcdecl.fbody = new CompoundStatement(Loc.initial, new Statements());
+    // funcdecl.isThis, funcdecl.isMember2
+    AggregateDeclaration ad, ad2;
 
-                if (funcdecl.naked)
-                {
-                    fpreinv = null;         // can't accommodate with no stack frame
-                    fpostinv = null;
-                }
+    uint oldErrors;
+    this(Scope* sc, FuncDeclaration funcdecl, uint oldErrors)
+    {
+        this.sc = sc;
+        this.funcdecl = funcdecl;
+        this.oldErrors = oldErrors;
+        needEnsure = FuncDeclaration.needsFensure(funcdecl);
+        f = cast(TypeFunction)funcdecl.type;
+    }
 
-                assert(funcdecl.type == f || (funcdecl.type.ty == Tfunction && f.purity == PURE.impure && (cast(TypeFunction)funcdecl.type).purity >= PURE.fwdref));
-                f = cast(TypeFunction)funcdecl.type;
+    void semantic()
```

Comment:
Naming this function `semantic` is really confusing. In dmd there are 3 semantic stages named bluntly dsymbolSemantic, semantic2 and semantic3. Although the function is defined in the file semantic3, on caller site it is not obvious to the reader which of the semantic passes it is calling. I would recommend renaming it `doSemantic3` or `semantic3Impl` or whatever that makes it obvious that it's the invocation of  semantic3